File: lmfetch/analyzers/importance.py
"""File importance scoring - ranks files by structural importance."""

from pathlib import Path

# Files that are typically entry points or important
HIGH_IMPORTANCE_PATTERNS = [
    "main.py", "app.py", "index.py", "server.py", "cli.py",
    "main.ts", "app.ts", "index.ts", "server.ts",
    "main.js", "app.js", "index.js", "server.js",
    "main.go", "main.rs", "lib.rs",
    "README.md", "readme.md", "README.rst",
    "setup.py", "pyproject.toml", "package.json", "Cargo.toml", "go.mod",
]

# Directories that typically contain important code
IMPORTANT_DIRS = [
    "src", "lib", "core", "api", "app", "server",
]

# Directories that typically contain less important code
LOW_IMPORTANCE_DIRS = [
    "test", "tests", "spec", "specs", "__tests__",
    "examples", "example", "samples", "sample",
    "docs", "doc", "documentation",
    "scripts", "tools", "utils", "helpers",
    "vendor", "third_party",
    "migrations", "fixtures",
]


def compute_file_importance(path: str) -> float:
    """Compute importance score for a file (0.0 to 1.0)."""
    p = Path(path)
    name = p.name.lower()
    parts = [part.lower() for part in p.parts]

    score = 0.5  # Base score

    # High importance files
    if name in [pat.lower() for pat in HIGH_IMPORTANCE_PATTERNS]:
        score += 0.3

    # Entry point patterns
    if name.startswith("main") or name.startswith("index") or name.startswith("app"):
        score += 0.15

    # __init__.py files are often important for understanding structure
    if name == "__init__.py" and len(p.parts) <= 3:
        score += 0.1

    # Important directories
    for important_dir in IMPORTANT_DIRS:
        if important_dir in parts:
            score += 0.1
            break

    # Low importance directories
    for low_dir in LOW_IMPORTANCE_DIRS:
        if low_dir in parts:
            score -= 0.2
            break

    # Deeper files are often less important
    depth = len(p.parts)
    if depth > 5:
        score -= 0.1 * (depth - 5)

    # Config/meta files
    if name.endswith(".json") or name.endswith(".yaml") or name.endswith(".yml"):
        if name not in ("package.json", "tsconfig.json", "pyproject.toml"):
            score -= 0.1

    return max(0.0, min(1.0, score))


def compute_centrality(
    path: str,
    dependency_graph: dict[str, set[str]],
) -> float:
    """Compute how central a file is based on imports (PageRank-like)."""
    if path not in dependency_graph:
        return 0.0

    # Count incoming edges (files that import this)
    incoming = sum(1 for deps in dependency_graph.values() if path in deps)

    # Count outgoing edges (files this imports)
    outgoing = len(dependency_graph.get(path, set()))

    # Normalize - files that are imported a lot are more central
    total_files = len(dependency_graph)
    if total_files == 0:
        return 0.0

    # Weighted: incoming matters more than outgoing
    centrality = (incoming * 2 + outgoing) / (total_files * 3)

    return min(1.0, centrality)

File: lmfetch-bun/src/analyzers/importance.ts
/**
 * File importance scoring heuristics
 */
import type { SourceFile } from "../sources/types";
import { basename, dirname } from "path";

// Entry point patterns (high importance)
const ENTRY_POINTS = new Set([
  "index.ts",
  "index.js",
  "index.tsx",
  "index.jsx",
  "main.ts",
  "main.js",
  "main.py",
  "mod.rs",
  "lib.rs",
  "main.rs",
  "main.go",
  "app.ts",
  "app.js",
  "app.py",
  "server.ts",
  "server.js",
  "cli.ts",
  "cli.js",
  "cli.py",
  "__init__.py",
  "setup.py",
  "package.json",
  "Cargo.toml",
  "go.mod",
]);

// Important directories (higher importance)
const IMPORTANT_DIRS = new Set([
  "src",
  "lib",
  "core",
  "api",
  "routes",
  "controllers",
  "services",
  "models",
  "components",
  "hooks",
  "utils",
  "helpers",
]);

// Less important directories (lower importance)
const LOW_IMPORTANCE_DIRS = new Set([
  "test",
  "tests",
  "__tests__",
  "spec",
  "specs",
  "e2e",
  "fixtures",
  "mocks",
  "stubs",
  "examples",
  "docs",
  "scripts",
  "tools",
  "config",
  "configs",
]);

// Less important file patterns
const LOW_IMPORTANCE_PATTERNS = [
  /\.test\./,
  /\.spec\./,
  /_test\./,
  /_spec\./,
  /\.d\.ts$/,
  /\.config\./,
  /\.mock\./,
];

export interface ImportanceScores {
  /** Map from file path to importance score (0-1) */
  scores: Map<string, number>;
}

/**
 * Calculate importance scores for all files
 */
export function calculateImportance(files: SourceFile[]): Map<string, number> {
  const scores = new Map<string, number>();

  for (const file of files) {
    const score = calculateFileImportance(file);
    scores.set(file.path, score);
    // Also store by relative path for easier lookup
    scores.set(file.relativePath, score);
  }

  return scores;
}

/**
 * Calculate importance score for a single file
 */
function calculateFileImportance(file: SourceFile): number {
  let score = 0.5; // Base score

  const fileName = basename(file.relativePath);
  const dirPath = dirname(file.relativePath);
  const dirs = dirPath.split("/").filter(Boolean);

  // Entry point bonus
  if (ENTRY_POINTS.has(fileName)) {
    score += 0.3;
  }

  // Important directory bonus
  for (const dir of dirs) {
    if (IMPORTANT_DIRS.has(dir)) {
      score += 0.1;
      break;
    }
  }

  // Low importance directory penalty
  for (const dir of dirs) {
    if (LOW_IMPORTANCE_DIRS.has(dir)) {
      score -= 0.2;
      break;
    }
  }

  // Low importance pattern penalty
  for (const pattern of LOW_IMPORTANCE_PATTERNS) {
    if (pattern.test(file.relativePath)) {
      score -= 0.15;
      break;
    }
  }

  // Depth penalty (deeper files are usually less important)
  const depth = dirs.length;
  if (depth > 3) {
    score -= (depth - 3) * 0.05;
  }

  // Root-level files bonus
  if (depth === 0) {
    score += 0.1;
  }

  // Language-based adjustments
  if (file.language === "markdown" || file.language === "mdx") {
    score -= 0.1;
  }
  if (file.language === "json" || file.language === "yaml") {
    score -= 0.05;
  }

  // Clamp to 0-1 range
  return Math.max(0, Math.min(1, score));
}

/**
 * Combine importance scores with centrality scores
 */
export function combineScores(
  importance: Map<string, number>,
  centrality: Map<string, number>,
  importanceWeight: number = 0.6
): Map<string, number> {
  const combined = new Map<string, number>();
  const centralityWeight = 1 - importanceWeight;

  // Get all unique paths
  const allPaths = new Set([...importance.keys(), ...centrality.keys()]);

  for (const path of allPaths) {
    const impScore = importance.get(path) ?? 0.5;
    const centScore = centrality.get(path) ?? 0.5;

    combined.set(path, impScore * importanceWeight + centScore * centralityWeight);
  }

  return combined;
}
